File: backend/app/schemas/funds.py
"""Schemas for unified funds/accounts view"""
from datetime import datetime
from typing import Dict, List, Optional
from pydantic import BaseModel, Field, ConfigDict
from app.models.funds import Institution, ConnectionStatus


class AccountBalance(BaseModel):
    """Latest balance snapshot for an account"""
    model_config = ConfigDict(from_attributes=True)

    account_id: int
    as_of: datetime
    current_balance: float
    available_balance: Optional[float] = None
    holdings_value: Optional[float] = None
    currency: str


class AccountWithBalance(BaseModel):
    """Account connection enriched with latest balance"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    institution: Institution
    account_name: str
    account_type: Optional[str] = None
    currency: str
    mask: Optional[str] = None
    status: ConnectionStatus
    last_synced_at: Optional[datetime] = None
    latest_balance: Optional[AccountBalance] = Field(
        default=None,
        description="Most recent balance snapshot for the account",
    )


class FundsSummary(BaseModel):
    """Aggregated balance view across all institutions"""
    totals_by_currency: Dict[str, float]
    last_synced_at: Optional[datetime]
    accounts: List[AccountWithBalance]
    fx_base_currency: Optional[str] = None
    fx_converted_total: Optional[float] = None


class FundsSummaryRequest(BaseModel):
    """Request body for FX-aware summary"""
    fx_base_currency: str = Field("USD", description="Target currency for conversion")
    fx_rates: Optional[Dict[str, float]] = Field(
        default=None,
        description="Map of currency code to rate in terms of target currency (e.g., CAD->USD rate)",
    )

File: frontend/nextjs/types/funds.ts
export type Institution = 'wealthsimple' | 'revolut' | 'ibkr' | 'td';
export type ConnectionStatus = 'connected' | 'requires_action' | 'error' | 'disconnected';

export interface AccountBalance {
  account_id: number;
  as_of: string;
  current_balance: number;
  available_balance?: number;
  holdings_value?: number;
  currency: string;
}

export interface AccountWithBalance {
  id: number;
  institution: Institution;
  account_name: string;
  account_type?: string;
  currency: string;
  mask?: string;
  status: ConnectionStatus;
  last_synced_at?: string;
  latest_balance?: AccountBalance;
}

export interface FundsSummary {
  totals_by_currency: Record<string, number>;
  last_synced_at?: string;
  accounts: AccountWithBalance[];
  fx_base_currency?: string;
  fx_converted_total?: number;
}

export interface FundsSummaryRequest {
  fx_base_currency: string;
  fx_rates?: Record<string, number>;
}

